### PR TITLE
Create an app for feed feature

### LINKED ISSUE #27: Missing Post Entity feature
A model that defines the Post needs to be created.
The Post represents a message posted by a user in the context of a specific course instructed by a tutor.

### PR DESCRIPTION
**Pull request by idan1356**
Creates the feed app along with its models and tests:

**models created:**
`Post`:  which represents a message posted by a user, in the context of a specific course instructed by a tutor.
`PostManager`, which is responsible for displaying hierarchized posts relevant for user by registered/own courses

**configure the project for said additions as follows:**
feed folder:
- Start app `feed`
- Create models `Post` and `PostManager` in `feed/models.py`
- Register models to `admin.py`
- Create initial migration file `0001_initial.py`

tests/feed folder:
-New fixtures and tests for new models

Resolves #27

### CODE REVIEW

**Review by shirazyomtov — CHANGES_REQUESTED** [review]
Nice work, I left some comments.
Also, I noticed that your PR didn't pass the flake8 automation.

File: feed/models.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    parent_post_id = models.ForeignKey("self", null=True, related_name='replies')
+    date = models.DateTimeField(auto_now_add=True)
+    content = models.TextField(null=False)
+    objects = PostManager()
+
+    # image = models.ImageField()
+
+    def is_root_post(self):
+        return not self.parent_post_id
+
```
[the remark this answers is not in the document]
Comment (idan1356) [reply feed/models.py:27]: You are correct, changed to `return self.parent_post_id is None`

File: feed/admin.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
+
```
[the remark this answers is not in the document]
Comment (idan1356) [reply feed/admin.py:2]: registered

**Comment by idan1356** [comment]
> Nice work, I left some comments. Also, I noticed that your PR didn't pass the flake8 automation.

Thanks. Regarding flake8 I'll make sure it passes before changing from draft to real PR.

**Review by shacharlevy — DISMISSED** [review]
LGTM!

**Review by shacharlevy — DISMISSED** [review]
LGTM!

File: feed/models.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        relevant_courses = TeacherCourses.objects.filter(studentcourses__in=relevant_users).distinct()
+        relevant_posts = Post.objects.filter(course_id__in=relevant_courses).order_by("date")
+
+        # create the post hierarchy in a dictionary that contains lists
+        post_hierarchy = {}
+        for post in relevant_posts:
+            if post.parent_post_id:
+                post_hierarchy[post.parent_post_id].append(post)
+            else:
+                post_hierarchy[post] = []
```
Comment (natifridman) [inline feed/models.py:24]: Maybe you can get this with a filter? something like you can do in SQL `WHERE parent_post_id==post`

File: feed/models.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+class Post(models.Model):
+    post_id = models.AutoField(primary_key=True)
+    course_id = models.ForeignKey(TeacherCourses, on_delete=models.CASCADE, related_name="related_posts")
+
+    user_id = models.ForeignKey(User, on_delete=models.CASCADE)
+    parent_post_id = models.ForeignKey("self", null=True, blank=True, on_delete=models.CASCADE)
+    date = models.DateTimeField(auto_now_add=True)
+    content = models.TextField(null=False)
+    # image = models.ImageField()
```
Comment (natifridman) [inline feed/models.py:37]: Is this in a comment on purpose?

File: feed/models.py (hunk 3 above)
Comment (idan1356) [reply feed/models.py:24]: I didn’t want to do additional queries, Would this option be preferable?
If so, I think I can access related replies by using something like `post.post__set`

File: feed/models.py (hunk 4 above)
Comment (idan1356) [reply feed/models.py:37]: Will uncomment

File: feed/models.py (hunk 3 above)
Comment (natifridman) [reply feed/models.py:24]: Yes, I meant to use something built-in instead of doing this for loop

Comment (natifridman) [reply feed/models.py:24]: I think you can do something like `relevant_posts.filter(parent_post_id= post.parent_post_id)` to get all the posts with this parent. I wonder if this functionality should be part of this method or somewhere else?

Comment (idan1356) [reply feed/models.py:24]: `relevant_posts` queryset contains all posts that are tied to user (posts that belong to a course the user is signed to/ owner of) regardless of whether post is a root post, or a reply post.

What I thought there is left to do is to organize the posts in a dictionary of lists where the key is a root post and the value is a list of replies (sorted by least recent to most recent reply). I thought this would be the most convenient data structure as it mimics how it should be displayed in the feed view.

maybe I still don't understand correctly what you mean,
but if I fetch for each post its replies using a filter it will lead to an additional query for each post (unless it gets optimized in a way I'm unaware of)

Comment (natifridman) [reply feed/models.py:24]: Got it, thanks

File: feed/models.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+from django.contrib.auth.models import User
+from course.models import TeacherCourses, StudentCourses
+
+
+class PostManager(models.Manager):
+    @staticmethod
+    def get_posts_for_user(user: User):
+        relevant_users = StudentCourses.objects.filter(
+            Q(Q(student_id=user) & Q(status="Confirmed") | Q(teacher_course_id__teacher_id=user))
```
Comment (Omeramsc) [inline feed/models.py:13]: Interesting use of Q!

File: feed/models.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+from django.db import models
+from django.core.exceptions import ValidationError
+from django.db.models import Q
+
+from django.contrib.auth.models import User
+from course.models import TeacherCourses, StudentCourses
+
+
+class PostManager(models.Manager):
```
Comment (Omeramsc) [inline feed/models.py:9]: I'm not against this class, but is it needed? what are the future plans for it? because I think `get_posts...` can be done in a class method for `Post`

File: tests/feed/test_feed.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+@pytest.mark.django_db
+class TestPost:
+    def test_is_root_post(self, persist_post_for_first_course, persist_reply_for_first_course):
+        assert persist_post_for_first_course.is_root_post()
+        assert not persist_reply_for_first_course.is_root_post()
+
+    def test_post_is_only_reply_to_root_post(self, persist_reply_for_first_course,
+                                             persist_post_for_second_course_student):
+        persist_post_for_second_course_student.parent_post_id = persist_reply_for_first_course
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:118]: If we are making a change to the fixture so early on, maybe we don't need a fixture

Comment (idan1356) [reply tests/feed/test_feed.py:118]: I think this comment I made explains my choice here:
https://github.com/redhat-beyond/HighTeach/pull/56#pullrequestreview-1393841803

File: feed/models.py (hunk 5 above)
Comment (idan1356) [reply feed/models.py:13]: What is the best practice for filtering based on a "nested" (not sure about the exact term) boolean expression?

File: feed/models.py (hunk 6 above)
Comment (idan1356) [reply feed/models.py:9]: I decided to separate the `get_posts...` method from the Post class and use a manager instead because:
1. This method is a "table level" query and not tied to a specific instance of Post

2. Relatively complex data operation compared to other queries inside the class methods.
Inside course docs for demo 2, it was defined that we need managers for:
`Manager classes for the more complex data operations you expect to do in your applications`

**Review by shacharlevy — APPROVED** [review]
LGTM!

File: feed/models.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            else:
+                post_hierarchy[post] = []
+
+        return post_hierarchy
+
+
+class Post(models.Model):
+    post_id = models.AutoField(primary_key=True)
+    course_id = models.ForeignKey(TeacherCourses, on_delete=models.CASCADE, related_name="related_posts")
+
```
[the remark this answers is not in the document]
Comment (idan1356) [reply feed/models.py]: done

File: feed/models.py (hunk 4 above)
Comment (idan1356) [reply feed/models.py]: done

File: feed/models.py (hunk 5 above)
Comment (Omeramsc) [reply feed/models.py]: I wasn't sarcastic, it is a nice usage. but if you found an error and fixed it, that's good :)

File: tests/feed/test_feed.py (hunk 7 above)
Comment (Omeramsc) [reply tests/feed/test_feed.py:130]: Personally I don't like that approach that much, because I expect a certain result of the test given the basic fixture, but then it is changed. I would rather, if needed, create another fixture that uses the current one, then using that new fixture instead. it can also have a readable name like "persist_reply_for_first_course_with..."

Comment (Omeramsc) [reply tests/feed/test_feed.py:130]: so It's bringing us back to the point of - maybe we don't need the fixtures. OR- maybe we need a `fixture factory`, to produce a fixture object on every test.

Comment (idan1356) [reply tests/feed/test_feed.py:130]: Ok got it!

Comment (idan1356) [reply tests/feed/test_feed.py]: fixed

**Review by shirazyomtov — APPROVED** [review]
LGTM

File: tests/feed/test_feed.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        with pytest.raises(ValidationError):
+            invalid_post.save()
+
+    def test_image_upload(self, image_file, persist_post_for_first_course):
+        persist_post_for_first_course.image = image_file
+        persist_post_for_first_course.save()
+        media_path = os.path.join(settings.MEDIA_ROOT, str(persist_post_for_first_course.image))
+
+        assert persist_post_for_first_course.image.read() == b"file_content"
+        assert os.path.isfile(media_path)
```
Comment (natifridman) [inline tests/feed/test_feed.py:153]: Why do you need this check? if it's saved in the DB, won't it be saved in `MEDIA_ROOT` for sure?

Comment (idan1356) [reply tests/feed/test_feed.py:153]: If i understand correctly images are not saved to DB using imageField, but to file system only,
I wanted to make sure I configured everything correctly and the file exists in media root after save.
 I guess the first check to read bytes is redundant

File: tests/feed/test_feed.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+    def test_is_user_able_to_post_in_course(self, persist_user, persist_course):
+        assert not Post.is_user_able_to_post_in_course(persist_user, persist_course)
+        user_joined_group = StudentCourses(student_id=persist_user, teacher_course_id=persist_course,
+                                           status="Confirmed")
+        user_joined_group.save()
+        assert Post.is_user_able_to_post_in_course(persist_user, persist_course)
+
+    def create_post_for_user_not_related_to_course(self, persist_user, persist_course):
+        assert not Post.is_user_able_to_post_in_course(persist_user, persist_course)
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:142]: what is invalid here? the post itself, or the fact the user is not related to the course? the test isn't very clear

File: tests/feed/test_feed.py (hunk 9 above)
Comment (Omeramsc) [reply tests/feed/test_feed.py:153]: Usually I recommend using URLs to images in those kind of projects as storing real images is complicated, but seems like you made it. good job

Comment (Omeramsc) [reply tests/feed/test_feed.py:153]: That being said, if you figured out how is it done, I'm not sure a test is needed, unless you check for specifics regarding the image saving functionality. then I'd think of a way to maybe separate this test from the Feed app, as it may be used in different parts of the whole application.

File: tests/feed/test_feed.py (hunk 10 above)
Comment (idan1356) [reply tests/feed/test_feed.py:142]: User should not be able to post in course because the user is not signed to course.
In this test, it is first asserted first that the user is not part of course, and then a post for said user for this course is attempted to be saved.

File: tests/feed/test_feed.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                               persist_first_student_course, persist_second_student_course,
+                               persist_post_for_first_course, persist_reply_for_first_course,
+                               persist_post_for_second_course_student):
+        expected = {persist_post_for_first_course: [persist_reply_for_first_course]}
+        posts_user = PostManager.get_posts_for_user(persist_user)
+
+        assert posts_user == expected
+        assert persist_post_for_second_course_student not in expected.items()
+
+    @pytest.mark.parametrize("num_of_posts", [4])
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:168]: Do we need a parameterized test for only one set of parameters? are there any other test cases we can use? maybe trying to fail it with edge cases?

also what are we testing? what order do we expect? either rename the test or use pytest.param(....) in the parametrization, and offer a name for each set of parameters describing what we want to achieve. 

e.g:
```
        pytest.param(
            "new_key",
            "new_value",
            id="test fail when new_key doesn't match new_value",
        ),
```

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Some things need changes, but overall impressive work

File: tests/feed/test_feed.py (hunk 9 above)
Comment (idan1356) [reply tests/feed/test_feed.py:153]: It does make more sense this test being outside the feed app tests, I'll move it to a different .py file inside the tests directory

Comment (idan1356) [reply tests/feed/test_feed.py:153]: Also, can it be part of this PR?

Comment (natifridman) [reply tests/feed/test_feed.py:153]: If it's relevant for this PR - yes.
If that makes the PR too big, maybe it should be in a different PR

**Review by Omeramsc — APPROVED** [review]
separate the image functionality/test to another PR and we will be able to merge this one

**Comment by idan1356** [comment]
> separate the image functionality/test to another PR and we will be able to merge this one

separated image functionality and tests from this PR

**Review by natifridman — APPROVED** [review]
LGTM

**Review by shirazyomtov — APPROVED** [review]
LGTM
